spec/controllers/users/phone_setup_controller_spec.rb

```diff
@@ -261,4 +263,54 @@
       end
     end
   end
+  describe 'check_phone_submission_limit' do
+    before do
+      @user = create(:user)
+      @user2 = create(:user)
+      @unconfirmed_phone = '+1 (202) 555-1213'
+    end
+    it 'rate limits use of phone by fingerprint' do
+      sign_in_before_2fa(@user)
+      allow(IdentityConfig.store).to receive(:otp_delivery_blocklist_maxretry).and_return(999)
+
+      freeze_time do
+        IdentityConfig.store.phone_submissions_per_fingerprint_limit.times do
+          post(:create, params: { new_phone_form: { phone: @unconfirmed_phone } })
+        end
+
+        expect(@user.reload.second_factor_locked_at).to eq Time.zone.now
+
+        timeout = distance_of_time_in_words(
+          RateLimiter.attempt_window_in_minutes(:phone_submissions).minutes,
+        )
+
+        expect(flash[:error]).to eq(
+          I18n.t(
+            'errors.messages.phone_confirmation_limited',
+            timeout: timeout,
+          ),
+        )
+        expect(response).to redirect_to account_path
+      end
+
+      sign_in_before_2fa(@user2)
+      freeze_time do
```

Comment:
Wouldn't it be important to ensure that signing in with the second user is also frozen in time, so that there's no race conditions as time elapses during the second user signing in?

Reply:
Yes. That's revised now.